src/mpid/ch4/src/ch4r_callbacks.h

```diff
@@ -114,14 +118,61 @@ static inline int MPIDI_handle_unexp_cmpl(MPIR_Request * rreq)
 
     if (MPIDI_CH4U_REQUEST(rreq, req->status) & MPIDI_CH4U_REQ_MATCHED) {
         match_req = (MPIR_Request *) MPIDI_CH4U_REQUEST(rreq, req->rreq.match_req);
+
+#ifndef MPIDI_CH4_DIRECT_NETMOD
+        if (unlikely(match_req && MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(match_req))) {
+            anysource_partner = MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(match_req);
+            if (!MPIR_STATUS_GET_CANCEL_BIT(anysource_partner->status)) {
+                mpi_errno = MPID_Cancel_recv(anysource_partner);
+                if (mpi_errno != MPI_SUCCESS) {
+                    goto fn_fail;
+                }
+                /* What should we do if the anysource partner request is not canceled? */
+                MPIR_Assertp(MPIR_STATUS_GET_CANCEL_BIT(anysource_partner->status));
```

Comment:
Not sure. This is an odd comment. I think it should be removed.

Reply:
I'm going through and adding a bunch of comments for this part of the code, but the gist of this part is this:

Unlike the case below, this should always be true because the request has already been matched. If the partner request can't be cancelled, then the request has been double-matched and we have a problem.